src/VisualStudio/Core/Def/Implementation/FindReferences/VisualStudioDefinitionsAndReferencesFactory.cs

```diff
@@ -50,25 +50,26 @@ internal class VisualStudioDefinitionsAndReferencesFactory
             if (result is not var (filePath, lineNumber, charOffset))
                 return null;
 
-            var displayParts = GetDisplayParts(filePath, lineNumber, charOffset);
+            await _threadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
+            var displayParts = GetDisplayParts_MustCallOnUIThread(filePath, lineNumber, charOffset);
             return new ExternalDefinitionItem(
                 definitionItem.Tags, displayParts,
                 _serviceProvider, _threadingContext,
                 filePath, lineNumber, charOffset);
         }
 
-        private ImmutableArray<TaggedText> GetDisplayParts(
+        private ImmutableArray<TaggedText> GetDisplayParts_MustCallOnUIThread(
```

Comment:
☹️ I would really prefer an attribute (similar to `[PerformanceSensitive]` for this), but we just don't have support for that yet.